subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/CrossProjectModelAccessTrackingParentDynamicObject.kt

```diff
@@ -0,0 +1,139 @@
+import org.gradle.configurationcache.problems.StructuredMessage
+import org.gradle.configurationcache.problems.location
+import org.gradle.internal.metaobject.DynamicInvokeResult
+import org.gradle.internal.metaobject.DynamicObject
+import java.util.Locale
+
+
+class CrossProjectModelAccessTrackingParentDynamicObject(
+    private val ownerProject: ProjectInternal,
+    private val delegate: DynamicObject,
+    private val referrerProject: ProjectInternal,
+    private val problems: ProblemsListener,
+    private val coupledProjectsListener: CoupledProjectsListener,
+    private val userCodeContext: UserCodeApplicationContext,
+) : DynamicObject {
+    override fun hasMethod(name: String?, vararg arguments: Any?): Boolean {
+        onAccess(MemberKind.METHOD, name)
+        return delegate.hasMethod(name, *arguments)
+    }
+
+    override fun tryInvokeMethod(name: String?, vararg arguments: Any?): DynamicInvokeResult {
+        onAccess(MemberKind.METHOD, name)
+        return delegate.tryInvokeMethod(name, *arguments)
+    }
+
+    override fun hasProperty(name: String?): Boolean {
+        onAccess(MemberKind.PROPERTY, name)
+        return delegate.hasProperty(name)
+    }
+
+    override fun tryGetProperty(name: String?): DynamicInvokeResult {
+        onAccess(MemberKind.PROPERTY, name)
+        return delegate.tryGetProperty(name)
+    }
+
+    override fun trySetProperty(name: String?, value: Any?): DynamicInvokeResult {
+        onAccess(MemberKind.PROPERTY, name)
+        return delegate.trySetProperty(name, value)
+    }
+
+    override fun getProperties(): MutableMap<String, *> {
+        onAccess(MemberKind.PROPERTY, null)
+        return delegate.properties
+    }
+
+    override fun getMissingProperty(name: String?): MissingPropertyException {
+        onAccess(MemberKind.PROPERTY, name)
+        return delegate.getMissingProperty(name)
+    }
+
+    override fun setMissingProperty(name: String?): MissingPropertyException {
+        onAccess(MemberKind.PROPERTY, name)
+        return delegate.setMissingProperty(name)
+    }
+
+    override fun methodMissingException(name: String?, vararg params: Any?): MissingMethodException {
+        onAccess(MemberKind.METHOD, name)
+        return delegate.methodMissingException(name, *params)
+    }
+
+    override fun getProperty(name: String?): Any {
+        onAccess(MemberKind.PROPERTY, name)
+        return delegate.getProperty(name)
+    }
+
+    override fun setProperty(name: String?, value: Any?) {
+        onAccess(MemberKind.PROPERTY, name)
+        return delegate.setProperty(name, value)
+    }
+
+    override fun invokeMethod(name: String?, vararg arguments: Any?): Any {
+        onAccess(MemberKind.METHOD, name)
+        return delegate.invokeMethod(name, *arguments)
+    }
+
+    private
+    enum class MemberKind {
+        PROPERTY, METHOD
+    }
+
+    private
+    fun onAccess(memberKind: MemberKind, memberName: String?) {
+        val trace = run {
```

Comment:
💄 I'd extract the whole `PropertyProblem` setup to its own method so this method can stay at a single level of abstraction:
```kotlin
    problems.onProblem(memberAccessProblem(memberKind, memberName))
    coupledProjectsListener.onProjectReference(referrerProject.owner, ownerProject.owner)
```

Reply:
Done in a way. Thanks!